[RISCV] Add scheduler definitions for XiangShan-KunMingHu

MrLinWang opened the pull request:
This patch adds an initial scheduler model for the XiangShan KunMingHu microarchitecture in the RISC-V backend, covering scalar instructions only. Support for vector instructions is not included in this patch.

The model includes initial ProcResource definitions and WriteRes mappings for scalar instructions, derived from the publicly available XiangShan KunMingHu documentation (including the [XiangShan Docs](https://docs.xiangshan.cc/projects/design/en/latest/) and [XiangShan User Guide](https://docs.xiangshan.cc/projects/user-guide/en/latest/)) as well as the [XiangShan KunMingHu RTL](https://github.com/OpenXiangShan/XiangShan) source code.

Simulator-based validation results will be provided in a follow-up update. Measurement results from real hardware will also be added afterwards to further validate and refine the model if necessary.

This implementation was contributed by the [XSCC Team](https://github.com/orgs/OpenXiangShan/teams/xscc).

MrLinWang committed:
[RISC-V] Add scheduler definitions for XiangShan-KunMingHu

MrLinWang committed:
Merge branch 'main' into OpenXiangShan/KunMingHu-SchedModel

MrLinWang committed:
Merge branch 'main' into OpenXiangShan/KunMingHu-SchedModel

MrLinWang committed:
[RISCV] Add XiangShan Kunminghu scheduling model and test cases

Add llvm-mca test cases for the XiangShan Kunminghu microarchitecture, including:
- Basic integer and floating-point operation tests
- Branch and memory access instruction tests
- Bit-manipulation extension tests
- Floating-point division and conversion tests

Also update the XiangShan Kunminghu scheduling model:
- Adjust pipeline resource groupings
- Optimize instruction latency and throughput
- Add support for bit-manipulation extensions
- Improve floating-point operation scheduling

MrLinWang committed:
Merge branch 'main' into OpenXiangShan/KunMingHu-SchedModel

MrLinWang committed:
[RISCV] Update XiangShan Kunminghu core documentation links

Update the XiangShan Kunminghu core documentation links to point to the latest official documentation and user guide.

Co-Authored-By: Chyaka <lilium23187@gmail.com>
Co-Authored-By: Bhe6669 <hebo@bosc.ac.cn>
Co-Authored-By: Tang Haojin <tanghaojin@outlook.com>
Co-authored-by: buggfg <wangyingying@bosc.ac.cn>

MrLinWang committed:
Merge branch 'main' into OpenXiangShan/KunMingHu-SchedModel

mshockwave commented on llvm/lib/Target/RISCV/RISCVSchedXiangShanKunMingHu.td:
```diff
@@ -0,0 +1,357 @@
+  def : WriteRes<WriteAtomicLDD, [XSPipeLDU]>;
+}
+
+let Latency = 1 in {
```
Latency is default to 1 already, so you can omit this line.

mshockwave commented on llvm/lib/Target/RISCV/RISCVSchedXiangShanKunMingHu.td:
```diff
@@ -0,0 +1,357 @@
+}
+
+/// Zb*
+let Latency = 1 in {
```
ditto

mshockwave commented on llvm/lib/Target/RISCV/RISCVSchedXiangShanKunMingHu.td:
```diff
@@ -0,0 +1,357 @@
+}
+
+// I2V
+let Latency = 0 in {
```
I thought "I2V" usually means "integer to vector".

Also, do you support move elimination across integer and floating point physical register files? where does this zero latency come from?

MrLinWang replied:
Thank you! The reason I used "I2V" is that, in the [Kunminghu Source Code in DecodeUnit](https://github.com/OpenXiangShan/XiangShan/blob/kunminghu-v2/src/main/scala/xiangshan/backend/decode/DecodeUnit.scala#L404)
, these instructions are defined as "I2V". However, I understand this could be misleading, as "I2V" is commonly associated with "integer to vector" conversions. To avoid confusion, I’ll update them to "I2F" as you suggested.

As for the zero latency, it’s defined in the [Kunminghu Source Code in I2vCfg Unit](https://github.com/OpenXiangShan/XiangShan/blob/kunminghu-v2/src/main/scala/xiangshan/backend/fu/FuConfig.scala#L273)
. I’ll check with the hardware team to confirm whether this value is appropriate.

Thanks again for your valuable input!

MrLinWang commented on llvm/lib/Target/RISCV/RISCVSchedXiangShanKunMingHu.td:
```diff
@@ -0,0 +1,357 @@
+// with a focus on achieving higher performance.
+// Source: https://github.com/OpenXiangShan/XiangShan
+// Documentation: https://docs.xiangshan.cc/projects/design/en/latest/
+// User Guide: https://docs.xiangshan.cc/projects/user-guide/en/latest/
```
Thanks very much for pointing this out and for the suggestion! We’re currently working with the XiangShan `KunMingHu` hardware team to publish more detailed microarchitecture documentation specific to `KunMingHu`. Once those materials become available, I’ll update this patch to reference them accordingly. This should help make the model easier to review and understand.

MrLinWang commented on llvm/lib/Target/RISCV/RISCVSchedXiangShanKunMingHu.td:
```diff
@@ -0,0 +1,357 @@
+// The latency of integer division ranges from 4 to 20.
+let Latency = 20, ReleaseAtCycles = [20] in {
+  def : WriteRes<WriteIDiv32, [XSPipeBJU3]>;
+  def : WriteRes<WriteIDiv, [XSPipeBJU3]>;
```
Good point, thanks for raising this! As mentioned above, once the detailed Kunminghu microarchitecture documentation becomes available, I’ll revisit the latency assignments in this scheduling model and update them accordingly. Thanks again for the suggestion!

MrLinWang committed:
Merge branch 'main' into OpenXiangShan/KunMingHu-SchedModel

MrLinWang committed:
[RISCV] Overhaul Kunminghu scheduling model and update MCA tests.

Co-Authored-By: Chyaka <lilium23187@gmail.com>
Co-Authored-By: Bhe6669 <hebo@bosc.ac.cn>
Co-Authored-By: Tang Haojin <tanghaojin@outlook.com>
Co-authored-by: buggfg <wangyingying@bosc.ac.cn>

MrLinWang committed:
Merge branch 'main' into OpenXiangShan/KunMingHu-SchedModel

MrLinWang committed:
[RISCV][llvm-mca] Fix total iteration count in XiangShan Kunminghu test checks

MrLinWang committed:
[RISCV] Remove unused LoopMicroOpBufferSize setting

MrLinWang committed:
[RISCV] NFC: Clean up KunMingHu scheduler comments

Remove trailing whitespace and stale I2V/F2V mentions from the KunMingHu scheduling model comments.

MrLinWang committed:
[RISCV] Fix doc URL path in XiangShan KunMingHu scheduling model

MrLinWang committed:
[RISCV] Update doc URL path in XiangShan KunMingHu scheduling model

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVSchedXiangShanKunMingHu.td:
```diff
@@ -0,0 +1,434 @@
+// Documentation: https://docs.xiangshan.cc/projects/design/en/latest/
+// User Guide: https://docs.xiangshan.cc/projects/user-guide/en/latest/
+// Typical Configurations & Instruction Latency:
+// https://docs.xiangshan.cc/projects/user-guide/en/
```
I don't know if we should follow the rule that we shouldn't exceed 80 characters in one line. Because for URL links, it is hard to copy/click if we break them.

MrLinWang replied:
I had the same question as well. The 80-character limit is useful for regular text, but URLs should perhaps be treated as an exception.

wangpc-pp replied:
I agree we shouldn't break URLs.

MrLinWang replied:
@topperc What do you think? I’d like to hear your opinion so we can reach a consistent formatting decision.

topperc replied:
I agree, we shouldn't break URLs.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVSchedXiangShanKunMingHu.td:
```diff
@@ -0,0 +1,434 @@
+
+def XSPipeBJU0 : ProcResource<1>; // BRU, JMP
+def XSPipeBJU1 : ProcResource<1>; // BRU, JMP
+def XSPipeBJU2 : ProcResource<1>; // BRU, JMP, I2F, VSet
```
Just curious about the design, why don't we merge BRU and ALU into one pipeline? What can we benefit from such fine-grained units?

MrLinWang replied:
The split here is mainly intended to match the actual hardware structure of the KunMingHu backend ([The block diagram of the structure of Kunminghu](https://docs.xiangshan.cc/projects/user-guide/en/kunminghu-v2/processor)).

Although ALU and BJU share the integer issue queue, they are different execution resources, with different supported functional units and resource conflicts. In the model, `XSPipeIntIQ*` represents the shared issue queue, while `XSPipeALU*` / `XSPipeBJU*` represent the different execution resources.

wangpc-pp reviewed (approved):
I did a quick look and I think it should be OK now!

LGTM!

MrLinWang committed:
Merge branch 'main' into OpenXiangShan/KunMingHu-SchedModel

MrLinWang committed:
文档(riscv): 更新典型配置URL为直接链接

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVSchedXiangShanKunMingHu.td:
```diff
@@ -0,0 +1,433 @@
+// Unsupported extensions
+defm : UnsupportedSchedQ;
+defm : UnsupportedSchedV;
+defm : UnsupportedSchedZfa;
```
I think XiangShan has supported `Zfa`.  Will you add it in later follow-ups?

MrLinWang replied:
Thank you for pointing this out. This PR focuses on the scheduling model for the instructions that are already prepared. Scheduling support for other extensions, including the V extension and Zfa, will be introduced in subsequent PRs.

MrLinWang committed:
Merge branch 'main' into OpenXiangShan/KunMingHu-SchedModel

MrLinWang committed:
Merge branch 'main' into OpenXiangShan/KunMingHu-SchedModel

MrLinWang committed:
Merge branch 'main' into OpenXiangShan/KunMingHu-SchedModel
